specs/electra/beacon-chain.md

```diff
@@ -1292,43 +1285,62 @@ def process_deposit_receipt(state: BeaconState, deposit_receipt: DepositReceipt)
+    # If the pending consolidations queue is full, consolidation requests are ignored
+    if len(state.pending_consolidations) == PENDING_CONSOLIDATIONS_LIMIT:
+        return
+    # If there is too little available consolidation churn limit, consolidation requests are ignored
+    if get_consolidation_churn_limit(state) <= MIN_ACTIVATION_BALANCE:
+        return
+
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    # Verify pubkeys exists
+    request_source_pubkey = execution_layer_consolidation_request.source_pubkey
+    request_target_pubkey = execution_layer_consolidation_request.target_pubkey
+    if request_source_pubkey not in validator_pubkeys:
+        return
+    if request_target_pubkey not in validator_pubkeys:
+        return
+    source_index = ValidatorIndex(validator_pubkeys.index(request_source_pubkey))
+    target_index = ValidatorIndex(validator_pubkeys.index(request_target_pubkey))
+    source_validator = state.validators[source_index]
+    target_validator = state.validators[target_index]
+
     # Verify that source != target, so a consolidation cannot be used as an exit.
-    assert consolidation.source_index != consolidation.target_index
+    if source_index == target_index:
+        return
+
+    # Verify source withdrawal credentials
+    has_correct_credential = has_execution_withdrawal_credential(source_validator)
+    is_correct_source_address = (
+        source_validator.withdrawal_credentials[12:] == execution_layer_consolidation_request.source_address
+    )
+    if not (has_correct_credential and is_correct_source_address):
+        return
+
+    # Verify that target has execution withdrawal credentials
+    if not has_execution_withdrawal_credential(target_validator):
+        return
 
-    source_validator = state.validators[consolidation.source_index]
-    target_validator = state.validators[consolidation.target_index]
     # Verify the source and the target are active
     current_epoch = get_current_epoch(state)
-    assert is_active_validator(source_validator, current_epoch)
-    assert is_active_validator(target_validator, current_epoch)
+    if not is_active_validator(source_validator, current_epoch):
+        return
+    if not is_active_validator(target_validator, current_epoch):
+        return
     # Verify exits for source and target have not been initiated
-    assert source_validator.exit_epoch == FAR_FUTURE_EPOCH
-    assert target_validator.exit_epoch == FAR_FUTURE_EPOCH
-    # Consolidations must specify an epoch when they become valid; they are not valid before then
-    assert current_epoch >= consolidation.epoch
-
-    # Verify the source and the target have Execution layer withdrawal credentials
-    assert has_execution_withdrawal_credential(source_validator)
-    assert has_execution_withdrawal_credential(target_validator)
-    # Verify the same withdrawal address
-    assert source_validator.withdrawal_credentials[12:] == target_validator.withdrawal_credentials[12:]
```

Comment:
@hwwhww it is great to see that the consolidation of validators with different withdrawal addresses is possible now.

Was there any reason why it was not allowed initially (why ever this constraint has existed?), but became possible by this PR?

Reply:
It was not allowed initially because initially consolidation was authorized by the validator’s pubkey which isn’t secure enough to change the owner (withdrawal creds) of the funds. In order to make this possible, a consolidation request system smart contract has been introduced on EL with the corresponding changes on CL to make it possible to sign requests with withdrawal creds which unlocked moving funds between validators with different creds

Reply:
Awesome. Appreciate the details 👍